src/csharp/Grpc.Core/Internal/ServerCallContextExtraData.cs

```diff
@@ -0,0 +1,97 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Grpc.Core.Internal
+{
+    /// <summary>
+    /// Additional state for <c>ServerCallContext</c>.
+    /// Storing the extra state outside of <c>ServerCallContext</c> allows it to be implementation-agnostic.
+    /// </summary>
+    internal class ServerCallContextExtraData
+    {
+        readonly CallSafeHandle callHandle;
+        readonly IServerResponseStream serverResponseStream;
+        readonly Lazy<AuthContext> cachedAuthContext;
+
+        public ServerCallContextExtraData(CallSafeHandle callHandle, IServerResponseStream serverResponseStream)
+        {
+            this.callHandle = callHandle;
+            this.serverResponseStream = serverResponseStream;
+            // TODO(jtattermusch): avoid unnecessary allocation of factory function and the lazy object.
+            this.cachedAuthContext = new Lazy<AuthContext>(GetAuthContextEager);
+        }
+
+        public ServerCallContext NewServerCallContext(ServerRpcNew newRpc, CancellationToken cancellationToken)
+        {
+            DateTime realtimeDeadline = newRpc.Deadline.ToClockType(ClockType.Realtime).ToDateTime();
+
+            return new ServerCallContext(this, newRpc.Method, newRpc.Host, realtimeDeadline,
+                newRpc.RequestMetadata, cancellationToken,
+                ServerCallContext_WriteHeadersFunc, ServerCallContext_WriteOptionsGetter, ServerCallContext_WriteOptionsSetter,
+                ServerCallContext_PeerGetter, ServerCallContext_AuthContextGetter, ServerCallContext_ContextPropagationTokenFactory);
+        }
+
+        private AuthContext GetAuthContextEager()
+        {
+            using (var authContextNative = callHandle.GetAuthContext())
+            {
+                return authContextNative.ToAuthContext();
+            }
+        }
+
+        // Implementors of ServerCallContext's members are pre-allocated to avoid unneccessary delegate allocations.
+        readonly static Func<ServerCallContext, object, Metadata, Task> ServerCallContext_WriteHeadersFunc = (ctx, extraData, headers) =>
+        {
+            return ((ServerCallContextExtraData)extraData).serverResponseStream.WriteResponseHeadersAsync(headers);
+        };
+
+        readonly static Func<ServerCallContext, object, WriteOptions> ServerCallContext_WriteOptionsGetter = (ctx, extraData) =>
+        {
+
+            return ((ServerCallContextExtraData)extraData).serverResponseStream.WriteOptions;
+        };
+
+        readonly static Action<ServerCallContext, object, WriteOptions> ServerCallContext_WriteOptionsSetter = (ctx, extraData, options) =>
```

Comment:
to add, I could imagine setting `ServerCallContext_WriteHeadersFunc` is an instance property within the constructor

Reply:
other option, might be less cumbersome paramer-wise if `ServerCallContext` had e.g. a `GetExtraData` method on it

Reply:
I considered both of these and decided not to do what you're suggesting:
- The extra "ServerCallContext" and "object" parameters in the lambdas is there to avoid unnecessary lambda allocations every time we create a ServerCallContext (the lambdas like ServerCallContext_WriteHeadersFunc need to be static.
- ServerCallContext.GetExtraData would have needed to be public and it would be visible to all the users writing the server side handlers - and I don't know that. If it was internal, every alternative grpc implementation using this API would need an [InternalsVisibleTo] annotation added to Grpc.Core.Api